src/utils.jl

```diff
@@ -616,23 +616,45 @@ macro refargs(ex)
     end
 end
 
-# https://github.com/JuliaLang/julia/issues/40626
-# as suggested in the above issue, spawned tasks may
-# end up getting stuck in thread local storage
-# running clear_thread_states clears out any thread local storage tasks
-struct _Returns{V} <: Function
-    value::V
-end
-
-(obj::_Returns)(@nospecialize(args...); @nospecialize(kw...)) = obj.value
-
-function clear_thread_states()
-    # only clear thread states in the workflow where a user is running
-    # CSV.File from the REPL at the top-level; if we're already spawned
-    # in a task, we don't want to mess up thread task states
-    if current_task() == Base.roottask
-        Threads.@threads :static for _ in 1:Threads.nthreads()
-            Timer(_Returns(nothing), 0; interval = 1)
-        end
+unwrap(val) = val
+unwrap(val::WeakRef) = val.value
+
+"""
+    @weakrefspawn args... begin
+
+    end
+
+A macro that wraps a `Threads.@spawn` block with `WeakRef` calls for all `args...` arguments, allowing
+them to be garbage collected once the `Task` has finished running. Must be used within a `@syncpreserve`
```

Comment:
Thank you for working on this!

Above `@sync` not `@syncpreserve` is used. Which is correct?

Also are we sure that the "preserve" part touches only the selected variables (this is relevant, as inside the task GC might have to be run if the taks itself allocates a lot).